addons/hr_expense/hr_expense.py

```diff
@@ -65,257 +32,219 @@ def _get_currency(self, cr, uid, context=None):
-                employee_payable_account_id = employee.address_home_id.property_account_payable.id
-        return {'value': {'department_id': department_id, 'company_id': company_id, 'employee_payable_account_id': employee_payable_account_id}}
-
-    def expense_confirm(self, cr, uid, ids, context=None):
-        for expense in self.browse(cr, uid, ids):
-            if not expense.line_ids:
+        return super(hr_expense_sheet, self).unlink()
+
+    @api.onchange('currency_id')
+    def onchange_currency_id(self):
+        account_journal = self.env['account.journal'].search([('type', '=', 'purchase'), ('currency', '=', self.currency_id.id), ('company_id', '=', self.company_id.id)], limit=1)
+        self.journal_id = account_journal.id
+
+    @api.multi
+    def sheet_draft(self):
+        for sheet in self:
+            sheet.line_ids.expense_line_draft()
+        return self.write({'state': 'draft'})
+
+    @api.multi
+    def sheet_confirm(self):
+        for sheet in self:
+            if not sheet.line_ids:
                 raise UserError(_('You cannot submit expense which has no expense line.'))
-            if expense.employee_id and expense.employee_id.parent_id.user_id:
-                self.message_subscribe_users(cr, uid, [expense.id], user_ids=[expense.employee_id.parent_id.user_id.id])
-        return self.write(cr, uid, ids, {'state': 'confirm', 'date_confirm': time.strftime('%Y-%m-%d')}, context=context)
-
-    def expense_accept(self, cr, uid, ids, context=None):
-        return self.write(cr, uid, ids, {'state': 'accepted', 'date_valid': time.strftime('%Y-%m-%d'), 'user_valid': uid}, context=context)
-
-    def expense_canceled(self, cr, uid, ids, context=None):
-        return self.write(cr, uid, ids, {'state': 'cancelled'}, context=context)
-
-    def account_move_get(self, cr, uid, expense_id, context=None):
+            if sheet.employee_id and sheet.employee_id.parent_id.user_id:
+                self.message_subscribe_users(user_ids = sheet.employee_id.parent_id.user_id.ids)
+            if False not in [True if expense.state in ['draft'] else False for expense in sheet.line_ids]:
+                sheet.line_ids.write({'state': 'confirm'})
+        return self.write({'date_confirm': fields.Date.context_today(self)})
+
+    @api.multi
+    def sheet_accept(self):
+        for expense in self.line_ids:
+            if expense.state != 'cancel':
+                expense.expense_approved()
+        return self.write({'state': 'accepted', 'date_valid': fields.Date.context_today(self), 'user_valid': self._uid})
+
+    @api.multi
+    def sheet_approved(self):
+        for sheet in self:
+            if sheet.state in ['accepted', 'done', 'paid']:
+                raise UserError(_('You have already approved sheet!'))
+            if sheet.state in ['cancelled']:
+                sheet.signal_workflow('draft')
+            sheet.signal_workflow('confirm')
+            sheet.signal_workflow('validate')
```

Comment:
@mdi-odoo
If expense sheet is in 'draft', 'confirm', 'cancel' state and we want to directly approve it from more menu then we have to manually trigger workflow signals.